src/model_theory/syntax.lean

```diff
@@ -378,6 +425,46 @@ def lift_at : ∀ {n : ℕ} (n' m : ℕ), L.bounded_formula α n → L.bounded_f
 | n (imp φ₁ φ₂) tf := (φ₁.subst tf).imp (φ₂.subst tf)
 | n (all φ) tf := (φ.subst tf).all
 
+/-- Sends a formula with constants to a formula with extra variables. -/
+def constants_to_vars : ∀ {n}, L[[γ]].bounded_formula α n → L.bounded_formula (γ ⊕ α) n
```

Comment:
Are we missing a general kind of functoriality here? Maybe something like:
to get a map `L1.bounded_formula a1 n -> L2.bounded_formula a2 n` we need a map `L1.term (a1 + b) -> L2.term (a2 + b)` for all `b` (or at least for `b = fin n` for all `n`) and we need `L1.rel n -> L2.rel n`. That is enough, right? Maybe we can generalize it even further to also incorporate the "make bound variables free" and "bind free variables" constructions...
I think there are many special cases of this construction already in the library: all constructions that don't do much with the

Reply:
and then you can prove functoriality (respecting compositions and identities) of this construction in general, and in particular that if the two given maps are equivalences, the resulting map is an equivalence.

Reply:
This is a good idea, I'm starting a new PR for that.

Reply:
#14466